indexing-service/src/main/java/org/apache/druid/indexing/worker/IntermediaryDataManager.java

```diff
@@ -0,0 +1,274 @@
+  // partitions.
+  // This can be null until IntermediaryDataManager is started.
+  @Nullable
+  private ScheduledExecutorService supervisorTaskChecker;
+
+  @Inject
+  public IntermediaryDataManager(
+      WorkerConfig workerConfig,
+      TaskConfig taskConfig,
+      IndexingServiceClient indexingServiceClient
+  )
+  {
+    this.intermediaryPartitionDiscoveryPeriodSec = workerConfig.getIntermediaryPartitionDiscoveryPeriodSec();
+    this.intermediaryPartitionCleanupPeriodSec = workerConfig.getIntermediaryPartitionCleanupPeriodSec();
+    this.intermediaryPartitionTimeout = workerConfig.getIntermediaryPartitionTimeout();
+    this.shuffleDataLocations = taskConfig
+        .getShuffleDataLocations()
+        .stream()
+        .map(config -> new StorageLocation(config.getPath(), config.getMaxSize(), config.getFreeSpacePercent()))
+        .collect(Collectors.toList());
+    this.indexingServiceClient = indexingServiceClient;
+  }
+
+  @LifecycleStart
+  public void start()
+  {
+    supervisorTaskChecker = Execs.scheduledSingleThreaded("intermediary-data-manager-%d");
+    // Discover partitions for new supervisorTasks
+    supervisorTaskChecker.scheduleAtFixedRate(
+        () -> {
+          for (StorageLocation location : shuffleDataLocations) {
+            final File[] dirsPerSupervisorTask = location.getPath().listFiles();
+            if (dirsPerSupervisorTask != null) {
+              for (File supervisorTaskDir : dirsPerSupervisorTask) {
+                final String supervisorTaskId = supervisorTaskDir.getName();
+                supervisorTaskCheckTimes.computeIfAbsent(
+                    supervisorTaskId,
+                    k -> DateTimes.nowUtc().plus(intermediaryPartitionTimeout)
+                );
+              }
+            }
+          }
+        },
+        intermediaryPartitionDiscoveryPeriodSec,
+        intermediaryPartitionDiscoveryPeriodSec,
+        TimeUnit.SECONDS
+    );
+    // Check supervisorTask status if its partitions have not been accessed in timeout.
+    // Delete partitions if the supervisorTask is already finished.
+    // Note that the overlord sends a cleanup request when a supervisorTask is finished. The below check is to trigger
+    // the self-cleanup for when middleManager misses the cleanup request from the overlord.
+    supervisorTaskChecker.scheduleAtFixedRate(
+        () -> {
+          final DateTime now = DateTimes.nowUtc();
+          final Set<String> expiredSupervisorTasks = new HashSet<>();
+          for (Entry<String, DateTime> entry : supervisorTaskCheckTimes.entrySet()) {
+            final String supervisorTaskId = entry.getKey();
+            final DateTime checkTime = entry.getValue();
+            if (checkTime.isAfter(now)) {
+              expiredSupervisorTasks.add(supervisorTaskId);
+            }
+          }
+
+          final Map<String, TaskStatus> taskStatuses = indexingServiceClient.getTaskStatuses(expiredSupervisorTasks);
```

Comment:
this has potential to throw exceptions and that would suppress further execution of this task... we should be stopping early only if the exception happened due to `interrupt`

Reply:
it is usually a good idea to put scheduled tasks inside a big try catch to avoid premature exit due to unknown exceptions.